### ISSUE TITLE
IngestExternalFile: question on usage

### ISSUE THREAD

**Issue by danhhz** [facebook/rocksdb#2473]
Hi, I'm working to use IngestExternalFile to power speedy ingest in https://github.com/cockroachdb/cockroach. As a result of some of the distributed consistency stuff, we have to handle retries. Things are much simpler if we can rely on it being safe to write an sstable, ingest it, write another _identical_ sstable, and then ingest that one. I can't see any reason why this wouldn't work and it's worked in my experiments, but I'd like to double check that this works the way I think it does.

Specifically:
1) I encode a sstable for an empty rocksdb keyrange and send it in an rpc
2) Some node receives that request, writes it to a file, and ingests it
3) That same node receives an identical request, writes it to a different file, and ingests it
4) A scan is run over the keyrange just ingested

I would expect the first file to ingest at the lowest level N. Then the second file would ingest at some higher level (probably N-1). The scan would return exactly the same thing as it would if the file had only been ingested once (because all of the keys and values are exactly the same in both of the files, so the iterator would always be returning the one in the second ingested file), though with some performance penalty for reading two files instead of one.

If I then run a compaction, both files would be merged into one with just a single copy of each of the imported keys (effectively ending up in exactly the same place as if we'd only ingested once).

Is this understanding correct?

cc @IslamAbdelRahman maybe

**Comment by mikhail-antonov** [comment]
IngestExternalFile will basically try to ingest file at the lowest possible level in the LSM  where the file "fits",  e.g. that this file will always override existing data associated with possible overlapping keyranges.

I actually don't think there's performance penalty to incur in this case as the reads will only touch the second file and never to go Ln to attempt to read first file.

I might be careful about 2 aspects here:

 - do you have interleaving writes to the DB? if yes they will be overridden by retries, obviously, so I assume that's not the case
 - Basically every time you run ingest on the same file it will need to be inserted at the higher and higher level in the LSM; how many levels do you have? At some point ingest files may start to go to L0 and cause read amplifications (also there's limit on the total number of L0 files).

On the more high level topic - Ingestion in RocksDB should be atomic in the sense that if you received an error that the file should have not been added to db; so I assume you are focusing on higher level (network) retries.

**Comment by IslamAbdelRahman** [comment]
Hi @danhhz

Yes your understanding is correct, the file number 2 will be ingested in a level higher than file number 1 and with a higher sequence number.

And yes there will be a read penalty if you are using an Iterator, but If you are doing a point look up, it wont matter because we wont check the file in Level N if we found our key in Level N -1

Can you please clarify more what is the problem you are trying to solve and what is your use case, It's not very clear to me.

**Comment by danhhz** [comment]
Thanks for the answers! Sorry it's taken me so long to respond.

Okay, that's probably enough to answer my question, but I'm happy to provide some additional background if you're interested.

CockroachDB uses raft to ensure consistency across replicas. This is simplifying things a lot, but when a raft command is applied, we have to apply the side effects (the actual data modification) as well as update a key which holds the id of the latest applied raft command. The issue motivating my question is crash protection. On each node, both the data and the applied raft key live in the same rocksdb instance, so we usually put them both in a rocksdb batch, which is atomic.

Unfortunately, IngestExternalFile is not available as an operation on a batch (and I assume this is not trivial to add), so we have to do the ingest (the side effect of a new raft command i'm adding) and update the applied index in serial. So we have to consider what happens if the process crashes in between these two ops.

If we write the applied index first and then the process crashes before the IngestExternalFile, then when it starts back up, it will think it already ingested it (and so it will never get ingested). This will mean that replica is inconsistent, which is not acceptable for us.

The other option is to call IngestExternalFile before updating the raft applied index. In this case, if the process crashes after the ingest and before the index update, then when it restarts, it will think the ingest wasn't called and redo it. This is what motivated my question above. If this is safe and produces the exact same behavior on iteration as if it had been ingested once, then we're set!

**Comment by gfosco** [comment]
Closing this via automation due to lack of activity.  If discussion is still needed here, please re-open or create a new/updated issue.
